### PR TITLE
Pass FileRequest's errors to the session (closes #1115)

### LINKED ISSUE #1115: Improve error handling in FileRequest
https://github.com/DevExpress/testcafe/issues/1434

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @churkin @LavrovArtem @miherlosev

### CODE REVIEW

**Comment by AndreyBelym** [comment]
FPR

File: src/request-pipeline/file-request.js [hunk 1]
```diff
@@ -2,34 +2,41 @@ import fs from 'fs';
...
         super();
 
-        var path = decodeURIComponent(parse(opts.url).pathname);
+        var parsed = parse(opts.url);
+        var path   = decodeURIComponent(parsed.pathname);
 
         if (DISK_RE.test(path))
             path = path.substr(1);
+        else if (parsed.hostname)
+            path = parsed.hostname + ':' + path;
```
Comment (miherlosev) [inline src/request-pipeline/file-request.js]: Do we need test for this condition?

File: src/utils/url.js [hunk 2]
```diff
@@ -290,7 +290,9 @@ export function prepareUrl (url) {
     // For example:
     // "//////example.com" -> "//example.com".
     // "http:///example.com" -> "http://example.com"
-    return url.replace(/^(https?:)?\/+(\/\/.*$)/i, '$1$2');
+    return url
+        .replace(/^(https?:)?\/+(\/\/.*$)/i, '$1$2')
+        .replace(/^file:\/*([A-Za-z]):/, 'file:///$1:');
```
Comment (miherlosev) [inline src/utils/url.js]: Add comment with example for `file://path`

File: src/request-pipeline/file-request.js [hunk 3]
```diff
@@ -2,34 +2,39 @@ import fs from 'fs';
...
+
 
 const DISK_RE = /^\/[A-Za-z]:/;
 
 export default class FileRequest extends EventEmitter {
     constructor (opts) {
         super();
 
-        var path = decodeURIComponent(parse(opts.url).pathname);
+        var parsed = parse(opts.url);
```
Comment (miherlosev) [inline src/request-pipeline/file-request.js]: Rename `parsed`  to `parsedUrl`

File: src/utils/url.js [hunk 4]
```diff
@@ -286,11 +286,14 @@ export function prepareUrl (url) {
     url = String(url);
     url = url.replace(/\n|\t/g, '');
 
-    // NOTE: Remove unnecessary slashes from the beginning of the url and after scheme.
+    // NOTE: Remove unnecessary or add missing slashes from the beginning of the url and after scheme.
```
Comment (miherlosev) [inline src/utils/url.js]: Let's split this comment on two sentences:
```js
// NOTE: Remove unnecessary slashes from the beginning of the url and after scheme.
 // "//////example.com" -> "//example.com".
// "http:///example.com" -> "http://example.com"
// "http:///example.com" -> "http://example.com"

// Or add missing slashes after the file scheme
// "file://C:\document.txt" -> "file:///C:\document.txt"
```

File: test/server/proxy-test.js [hunk 5]
```diff
@@ -1076,6 +1076,31 @@ describe('Proxy', function () {
             });
         });
 
+        it('Should process page with non-conforming Windows url', function (done) {
```
Comment (miherlosev) [inline test/server/proxy-test.js]: Move out condition from the test:
```js
if (os.platform() === 'win32') {
    it('Should process page with non-conforming Windows url', function (done) {}
}
```

File: src/utils/url.js [hunk 6]
```diff
@@ -286,11 +286,16 @@ export function prepareUrl (url) {
     url = String(url);
     url = url.replace(/\n|\t/g, '');
 
-    // NOTE: Remove unnecessary slashes from the beginning of the url and after scheme.
+    // NOTE: Remove unnecessary from the beginning of the url and after scheme.
```
Comment (miherlosev) [inline src/utils/url.js]: You lost the `slashes` word.

File: test/server/proxy-test.js [hunk 7]
```diff
@@ -1076,6 +1076,28 @@ describe('Proxy', function () {
             });
         });
 
+        if (os.platform() !== 'win32') {
```
Comment (miherlosev) [inline test/server/proxy-test.js]: `os.platform() === 'win32'`

**Pull request merged**